### PR TITLE
[Core] Expanded job ID to 4 bytes by removing object flag bytes.

### LINKED ISSUE #8032: [Core] Recycle driver job IDs.
### Describe your feature request

For a long-lived Ray cluster with many (not necessarily concurrent) drivers, the simple job ID provisioning of consulting GCS to receive an incremented job ID eventually causes a returned job ID to exceed the maximum `JobID`.

```
  File "python/ray/worker.py", line 1155, in connect
    int(worker.redis_client.incr("JobCounter")))
  File "python/ray/includes/unique_ids.pxi", line 292, in ray._raylet.JobID.from_int
AssertionError: Maximum JobID integer is 65535.
```

It'd be nice if these auto-generated job IDs could be recycled, so instead of being limited to 65535 jobs/drivers for the lifetime of the Ray cluster, you'd be limited to 65535 _concurrent_ jobs/drivers, which users will probably be much more hard-pressed to hit.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Expanded the job ID to 4 bytes by removing the flag bytes from the object ID, in an effort to support more than 65535 Ray jobs in the lifetime of a Ray cluster (see [issue](https://github.com/ray-project/ray/issues/8032)). Although the `created_by_task` and `object_type` flags are set, they don't appear to be used by any internal consumers, and methods for accessing those flags aren't exposed to the client.

I initially decided to not consolidate `ForPut` and `ForTaskReturn`, despite their functionality being exactly the same, primarily in an effort to reduce higher-level API impact and to make the scope of this PR small. Although the semantics of these functions haven't changed, merging these functions into a single `ForTask` (or similarly named) function and updating the callers is probably a better design.

## Why are these changes needed?

To support more than 65535 Ray jobs in the lifetime of a Ray cluster.

## Related issue number

Closes #8032 

## TODOs

- [x] Update [ID specification](https://github.com/ray-project/ray/blob/d3bac298d5af73df0a1fc71d5e8f0c598dd21342/src/ray/design_docs/id_specification.md).
- [x] Add comment to worker context explaining the need to initialize the `put_index_` counter with the number of returns for the current task.
- [x] Consider consolidating `ForPut` and `ForTaskReturn`, or decide to put off to a future PR.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by edoakes** [comment]
Thanks @clarkzinzow, this looks great. There seems to be a test failing due to an unused variable:
https://travis-ci.com/github/ray-project/ray/jobs/375156693#L840

File: src/ray/common/id.cc [hunk 1]
```diff
@@ -259,42 +203,28 @@ TaskID ObjectID::TaskId() const {
 ObjectID ObjectID::ForPut(const TaskID &task_id, ObjectIDIndexType put_index) {
```
Comment (edoakes) [inline src/ray/common/id.cc:203]: As you mentioned in the description I think it'd be best to collapse these into one definition for clarify - something like `ObjectID::FromIndex`

File: src/ray/core_worker/context.cc [hunk 2]
```diff
@@ -56,7 +57,8 @@ struct WorkerThreadContext {
   /// Number of tasks that have been submitted from current task.
   int task_index_;
 
-  /// Number of objects that have been put from current task.
+  /// Number of return objects for the current task plus the number of objects that
+  /// have been put from current task.
```
Comment (edoakes) [inline src/ray/core_worker/context.cc:61]:
```suggestion
  /// Used to calculate ObjectIDs for put objects. The index starts at the number of return values from this task
  /// to reserve those indices for the return objects.
```

**Review by edoakes — COMMENTED** [review]
This looks great! Could you make sure to document this clearly everywhere that it's relevant? I'm a little bit concerned that it's very "clever" which might make peoples' job harder in the future if they don't understand that we reserve the first few indices for the return objects.

At the very least, let's make sure to update `id_specification.md`

File: src/ray/common/id.cc (hunk 1 above)
Comment (clarkzinzow) [reply src/ray/common/id.cc:203]: Definitely. What do you think about `FromTaskIndex`? I feel like that pretty clearly conveys the semantics, not sure if including `Task` is redundant.

File: src/ray/core_worker/context.cc (hunk 2 above)
Comment (clarkzinzow) [reply src/ray/core_worker/context.cc:61]: Way better, thanks. This is almost word-for-word what I was going to update it to. 😁

File: src/ray/common/id.cc (hunk 1 above)
Comment (edoakes) [reply src/ray/common/id.cc:203]: Preference for `FromIndex` - `Task` is redundant and this can either be a `PutIndex` or a `ReturnIndex`

**Comment by rkooo567** [comment]
@clarkzinzow Just lyk our master build has been stabilized a lot compared to the past, and if you don't pass all 3 builds on the top, it is likely that there are issues from the PR!

File: src/ray/core_worker/context.cc [hunk 3]
```diff
@@ -56,7 +59,10 @@ struct WorkerThreadContext {
   /// Number of tasks that have been submitted from current task.
   int task_index_;
 
-  /// Number of objects that have been put from current task.
+  /// Used to calculate ObjectIDs for put objects. The index starts at the number of
+  /// return values for the current task in order to keep the put indices from
+  /// conflicting with return object indices. 0 < idx <= NumReturns() is reserved for
+  /// return objects, while idx > NumReturns is available for put objects.
```
Comment (edoakes) [inline src/ray/core_worker/context.cc:65]: Nice!

File: src/ray/design_docs/id_specification.md [hunk 4]
```diff
@@ -49,22 +49,6 @@ Note: Dummy actor id is an `ActorID` whose unique part is nil.
 ```
 
 #### ObjectID (20 bytes)
-An `ObjectID` contains 3 parts:
+An `ObjectID` contains 2 parts:
 - `index bytes`: 4 bytes to indicate the index of the object.
```
Comment (edoakes) [inline src/ray/design_docs/id_specification.md:53]: Comment here about the task returns?

**Review by edoakes — APPROVED** [review]
LGTM

**Comment by edoakes** [comment]
Looks like there are some relevant failures in CI, e.g., https://travis-ci.com/github/ray-project/ray/jobs/375554361#L5943

**Pull request merged**